sqlmesh/core/engine_adapter/base.py

```diff
@@ -67,6 +68,32 @@ def is_insert_overwrite(self) -> bool:
     def is_replace_where(self) -> bool:
         return self == InsertOverwriteStrategy.REPLACE_WHERE
 
+    @property
+    def requires_condition(self) -> bool:
+        return self.is_replace_where or self.is_delete_insert
+
+
+@dataclass
+class SourceQuery:
+    generator: t.Iterator[Query]
+    columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None
+    batched: bool = False
+    is_from_df: bool = False
+
+    @property
+    def single_query(self) -> t.Iterator[Query]:
```

Comment:
If it's a single query, why yield?

Reply:
I'm guessing you wanted to make it explicit that this property cannot be re-entered just like `generator`, but it's not obvious at all from either the method name or the signature. I'd rather just cache the query instead and always return it.

Reply:
why does it matter if this method can be re-entered or not?

Reply:
Some engine adapter methods have explicit logic for if they have X number of queries or a single query. For example creating a view on a pandas dataframe wants to assert that it can't get more than a single query back. If we just returned a query though we would loose the lazy evaluation property of the generator which allows the provider of the generator to say how things should behave pre/post the creation of the query. 

For why does it matter if we re-enter, some engines create temporary tables for dataframes and therefore if we re-enter then we would waste resources creating a redundant set of temporary tables.

Reply:
> why does it matter if this method can be re-entered or not?

This means that now at any given place where the `SourceQuery` is used I have to guess in which state it's currently in and whether its generator is valid or not. Provided that the `SourceQuery` is not properly scoped and can be used in a multitude of implementations downstream, I think its stateful nature adds unnecessary complexity.

Reply:
IMHO another alternative is to let `SourceQuery` encapsulate a process that **produces** a generator and expose a method called `run` or `generate` that creates the generator and yields results rather than caching the generator itself:
```
class SourceQuery:
    generator_factory: t.Callable[[], t.Iterator[Query]]
    
    def generate(self) -> t.Iterator[Query]:
        yield from generator_factory()
```
This way the object is no longer stateful and the `generate` method can be safely re-entered. So regardless of where and when I call it it'd always produce the same result. We can of course cache individual items if needed and this will introduce state, but at least it will be well-scoped, encapsulated and safe.

Reply:
Generally though, if I understood the idea correctly, I don’t understand why we should couple something stateless like query generation with side effects like creation of temporary tables. I’d rather make side effects explicit than present them as generic iterator.

Reply:
>  I don’t understand why we should couple something stateless like query generation with side effects like creation of temporary tables

The query is based on the table generated. So temporary table `temp_123` is created and then the query is `select cola, colb from temp_123`. So they are technically coupled. And then destroying the table is coupled to the query to because we only want to drop the temporary table once it is no longer needed and that is once the query has been executed. This is what the current generator provides out-of-box with a contract on callers that yielding queries can only be done once (or run the risk of wasting resources).

I could try to do something where I create assets based on a query and then it could be:
Generate Query -> Process Query with Pre-hook -> Yield Query to caller -> Process query with Post-hook. This would mean you could yield queries from the generator freely and skip hooks if you just want to inspect the queries themselves. Example code (building on what you have above):
```
class SourceQuery:
    generator_factory: t.Callable[[], t.Iterator[Query]]
    pre_hook: t.Optional[t.Callable[[Query], None]] = None
    post_hook: t.Optional[t.Callable[[Query], None]] = None
    columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None

    def generate(self) -> t.Iterator[Query]:
        yield from self.generator_factory()

    @property
    def run(self) -> t.Iterator[Query]:
        for query in self.generate():
            if self.pre_hook:
                self.pre_hook(query)
            yield query
            if self.post_hook:
                self.post_hook(query)
```

Note that this is capturing exactly what is happening currently. It though makes this pre/post hook behavior more explicit and provides an interface for iterating through all objects without doing that extra processing. Note that this is only helpful when creating the query itself can be done cheaply and it assumes all the pre/post hooks need is a query to do their work.

Reply:
you should just implement the iterator in the class, make it an actual iterator / generator

Reply:
Talked and this is the follow up this: https://github.com/TobikoData/sqlmesh/pull/1362/commits/e7892aa9d51179c07d7c53133ac8bf318809b39a